pkg/secretlock/local/local_secret_lock.go

```diff
@@ -0,0 +1,176 @@
+
+// Lock is a secret lock service responsible for encrypting keys using a master key
+type Lock struct {
+	aead cipher.AEAD
+}
+
+// MasterKeyFromPath creates a new instance of a local secret lock Reader for a master key stored in `path`.
+func MasterKeyFromPath(path string) (io.Reader, error) {
+	masterKeyFile, err := os.OpenFile(path, os.O_RDONLY, 0600)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		err = masterKeyFile.Close()
+		if err != nil {
+			logger.Warnf("failed to close file: %w", err)
+		}
+	}()
+
+	mkData := make([]byte, 1024)
+
+	n, err := io.ReadFull(masterKeyFile, mkData)
+	if err != io.ErrUnexpectedEOF {
+		return nil, err
+	}
+
+	mkData = mkData[0:n]
+
+	return bytes.NewReader(mkData), nil
+}
+
+// MasterKeyFromEnv creates a new instance of a local secret lock Reader
+// for a master key found in a env variable with key: `envPrefix` + `keyURI`.
+func MasterKeyFromEnv(envPrefix, keyURI string) (io.Reader, error) {
+	mk := os.Getenv(envPrefix + strings.ReplaceAll(keyURI, "/", "_"))
+	if mk == "" {
+		return nil, fmt.Errorf("masterKey not set")
+	}
+
+	return bytes.NewReader([]byte(mk)), nil
+}
+
+// NewService creates a new instance of local secret lock service using a master key in masterKeyReader.
+func NewService(masterKeyReader io.Reader, secLock secretlock.Service) (secretlock.Service, error) {
+	masterKeyData := make([]byte, 1024)
+
+	if masterKeyReader == nil {
+		return nil, fmt.Errorf("masterKeyReader is nil")
+	}
+
+	n, err := masterKeyReader.Read(masterKeyData)
+	if err != nil {
+		if err != io.EOF && err != io.ErrUnexpectedEOF {
+			return nil, err
+		}
+	}
+
+	if n == 0 {
+		return nil, fmt.Errorf("masterKeyReader is empty")
+	}
+
+	masterKey := make([]byte, n)
+
+	// copy masterKey read from reader directly
+	copy(masterKey, masterKeyData)
+
+	// if secLock not empty, then masterKeyData is encrypt, let's decrypt it first.
+	if secLock != nil {
+		decResponse, e := secLock.Decrypt("", &secretlock.DecryptRequest{
+			Ciphertext: string(masterKeyData[0:n])})
+		if e != nil {
+			return nil, e
+		}
+
+		masterKey = []byte(decResponse.Plaintext)
+	}
+
+	aead, err := createAESCipher(masterKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Lock{aead: aead}, nil
+}
+
+// Encrypt a key in req using master key in the local secret lock service
+// (keyURI is used for remote locks, it is ignored by this implementation)
+func (s *Lock) Encrypt(keyURI string, req *secretlock.EncryptRequest) (*secretlock.EncryptResponse, error) {
+	nonce := random.GetRandomBytes(uint32(s.aead.NonceSize()))
+	ct := s.aead.Seal(nil, nonce, []byte(req.Plaintext), []byte(req.AdditionalAuthenticatedData))
+	ct = append(nonce, ct...)
+
+	return &secretlock.EncryptResponse{
+		Ciphertext: base64.URLEncoding.EncodeToString(ct),
+	}, nil
+}
+
+// Decrypt a key in req using master key in the local secret lock service
+// (keyURI is used for remote locks, it is ignored by this implementation)
+// nolint:dupl
```

Comment:
what is being duplicated?

Reply:
this implementation function has the same logic in `Lock` and `masterLockHKDF`, although both have their own instance of aead cipher.. the linter still complains the functions are duplicates.

Reply:
I'll probably move masterLockHKDF into a sub package

Reply:
TODO - extract common code into a function

Reply:
Decrypt() functions are in different packages now.. removed the `nolint:dupl` comment